In-app feedback with Nolt.io

AWolf81 opened the pull request:
**Related Issue:**
#243

**Summary:**
Added [Nolt.io](https://nolt.io). It opens a new Electron window from help/Feedback or from the UI.
If the user minimizes the second screen and clicks the the feedback button again. It won't open a new feedback screen. It displays the same window.

**Todos:**
- [ ] Change Nolt board to a board from @joshwcomeau (if it's OK to use Nolt)

**Things to discuss:**
- Feedback button position in lower right corner OK? Or should we place it at a different location
- Feedback button styling OK? Coloring like the settings button.
- Is it OK to use Nolt? I haven't used it before. I've googled a lot and that's the only service I've found that is free. But there are probably other good services. I like Nolt as styling is a bit customizable and also integrates Github (not sure how good the Github integration is but it would be great if we could create new issues from a feedback post).
- It would be nice to display the most recent posts/comments on hover of the feedback button. Not sure if this is possible - I have to check the API of Nolt.
- Display a number on the feedback button if there are new posts/comments (if possible)
- Do we have to update privacy policy? I think we should add a note that Nolt is a data processor with a link to their privacy policy.

**Screenshots/GIFs:**
*Application menu*
![grafik](https://user-images.githubusercontent.com/3046542/48968829-798b5300-eff5-11e8-8f9c-b2adc2503d59.png)

*New window*
![grafik](https://user-images.githubusercontent.com/3046542/48968794-08e43680-eff5-11e8-9a93-c9594ab45f5a.png)

*Feedback button*
![screenrecording_feedback_button](https://user-images.githubusercontent.com/3046542/48968802-21ece780-eff5-11e8-9f4e-c9ca9dfa8362.gif)

melanieseltzer commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,73 @@
+
+class FeedbackButton extends PureComponent<Props> {
+  static defaultProps = {
+    size: 36,
```
```suggestion
    size: 20,
```

(Part of overall UI suggestion)

melanieseltzer commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,73 @@
+class FeedbackButton extends PureComponent<Props> {
+  static defaultProps = {
+    size: 36,
+    color: COLORS.gray[400],
```
```suggestion
    color: COLORS.transparentWhite[300],
```

(Part of overall UI suggestion)

melanieseltzer commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,73 @@
+  static defaultProps = {
+    size: 36,
+    color: COLORS.gray[400],
+    hoverColor: COLORS.purple[500],
```
```suggestion
    hoverColor: COLORS.white,
```

(Part of overall UI suggestion)

melanieseltzer commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,73 @@
+          <Spring
+            native
+            to={{
+              scale: isHovered ? 1.15 : 1,
```
```suggestion
              scale: isHovered ? 1.05 : 1,
```

(Part of overall UI suggestion)

melanieseltzer commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,73 @@
+            }}
+          >
+            {({ scale }) => (
+              <IconWrapper scale={scale} color={isHovered ? hoverColor : color}>
```
```suggestion
              <IconWrapper scale={scale} color={isHovered ? hoverColor : color} onClick={() => openWindow(IN_APP_FEEDBACK_URL)}>
```

(Part of overall UI suggestion - remove onClick from `<IconBase>` and put here instead)

melanieseltzer commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,73 @@
+      transform: `scale(${scale}, ${scale})`,
+    }),
+  })`
+    color: ${props => props.color};
```
```suggestion
    color: ${props => props.color};
    cursor: pointer;
    background: linear-gradient(45deg, ${COLORS.violet[700]}, ${COLORS.purple[500]});
    width: auto;
    border-radius: 20px;
    display: flex;
    align-items: center;
    padding: 10px;
    font-size: 16px;
```

(Part of overall UI suggestion)

melanieseltzer commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,73 @@
+
+const DetectActiveWrapper = styled(DetectActive)`
+  position: fixed;
+  width: 40px;
```
```suggestion
  
```

(Part of overall UI suggestion - remove width here)

melanieseltzer commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,73 @@
+    color: ${props => props.color};
+  `
+);
+
```
```suggestion
const IconBaseWrapper = styled(IconBase)`
  margin-right: 5px;
`;
```

(Part of overall UI suggestion)

melanieseltzer commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,73 @@
+          >
+            {({ scale }) => (
+              <IconWrapper scale={scale} color={isHovered ? hoverColor : color}>
+                <IconBase
```
```suggestion
                <IconBaseWrapper
```

(Part of overall UI suggestion - rename for styled-component)

melanieseltzer commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,73 @@
+import styled from 'styled-components';
+import IconBase from 'react-icons-kit';
+import { messageSquare } from 'react-icons-kit/feather/messageSquare';
+import { Spring, animated, interpolate } from 'react-spring';
```
nit: interpolate is unused

AWolf81 commented:
@melanieseltzer Thanks for your review. I'll look into it later today.

I like the minimal design of just the icon but you're right a text would be good. But I won't add it to the UI. Maybe just adding a tooltip would be better. So by hovering the user knows what it will do.

Also with the small icon I think it's not that critical if it's above other controls.

No, there was no decision against Doorbell. I just decided against it as there is no free plan. The free is just a 7 day free trial.
I would start with Nolt. I think it's also no problem to switch this later to a differnet service if there is a better one but I think Nolt is exactly what we need.

j-f1 commented:
How about sliding the text out on hover? Google Docs does something similar with the explore button.

melanieseltzer commented:
> No, there was no decision against Doorbell. I just decided against it as there is no free plan. The free is just a 7 day free trial.

Ah makes sense. Yeah Nolt seems decent enough.

> How about sliding the text out on hover? Google Docs does something similar with the explore button.

I like this a lot ^

AWolf81 commented:
I also like the slide out feedback text. I'll add it to Guppy later today.

[Here](https://codesandbox.io/s/lrjp828yxq) is a demo of the button that I'll add. I like it as it is. Should I improve anything?
I wanted to have some transparent parts but that wasn't working. The icon background needs to be a solid color as it hides the text.
The slider could be a bit transparent but with text below it looks weird.

AWolf81 commented:
The updated animation should be OK. See updated screenrecording.

I've also changed the opening of Nolt to open in browser as I think that's better - see todo list item about a tweet with a security concern about login in a desktop app. I've kept the code for `windowOpen`  so we could re-add that if it's OK to have a login in app.
I think it would be OK to open in Guppy as the user have to trust Guppy as it can do anything on the computer - so by starting the executable the user needs to be sure that he downloaded Guppy from the right place (with-out malicious code etc.). Maybe a Virus check & checking the MD5/SHA256 would help to check that it's safe to execute - we should add these checksums to future releases.

@joshwcomeau I think just switching Nolt board to a board that you've created is open. If you're OK with Nolt, can you please create a new board and let me know the url?
We also need to create some starting posts - similar to what I have added in the example board.

idoberko2 commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,157 @@
+        }}
+        to={{ hovered: isHovered ? 1.0 : 0 }}
+      >
+        {({ hovered }) => (
```
@AWolf81, In order to avoid props that look like `hovered={hovered}`, @joshwcomeau suggested to not destructure the arguments of the child function, and I widely adopted this convention in my branch. I usually named the argument `interpolated` and then the prop of the child component would be `hover={interpolated.hover}`.
Other than that, it looks good to me.

AWolf81 replied:
@idoberko2 thanks for reviewing.
OK, I think that's more a "style" guide thing and it looks better if not destructured. It's also more clear that the hovered value passed to the styled component is the interpolated value from React-Spring.
I'll change that.

We'll add this to our docs so it's clear that we're doing it like this. @joshwcomeau should we add it as a note to issue #210 - so we're having a reminder?

AWolf81 commented:
@melanieseltzer no problem, thanks for reviewing. 
I'm busy with the `export-to-codesandbox` feature but I did some progress locally (converted logic to a saga & added unit tests).  Will push a update soon.

I've tried hard to get the rounded border from the beginning but I haven't found a solution. It always scaled the radius.
If we find a way to get the radius from the beginning it's OK to change. I think we could also improve the spring speed as during leave animation the slider is almost faster than the box - that's also better in the demo as this couldn't happen there as the slider & text were tied to each other.

The demo is animating the padding and I wanted to only use scale & transition to get a good performance. I think the performance of the demo is OK but just using transform will be better.

Yes, this is ready to be merged. We just have to wait for the feedback from @joshwcomeau as he should create the Nolt board. Then I'll change the link to that board.

joshwcomeau commented on src/components/ApplicationMenu/ApplicationMenu.js:
```diff
@@ -60,6 +61,10 @@ class ApplicationMenu extends Component<Props> {
 
+  openLink = url => {
+    shell.openExternal(url);
+  };
```
Nit: Personal opinion is that this abstraction doesn't add a ton of value =) I'd just call `shell.openExternal` directly.

But yeah that's a personal preference thing, so no changes necessary.

AWolf81 replied:
yes, you're right. That's not needed. I'll change it and inline it.

joshwcomeau commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,160 @@
+  width: 50px;
+  height: 50px;
+  text-align: center;
+  z-index: +1;
```
Nit: + isn't necessary (unless this is syntax I'm not familiar with?)

joshwcomeau commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,160 @@
+  height: 50px;
+  /*
+  // would be nice to have a border-radius but scale is modifying it
+  // so it's looking weird - I think it's OK to just have the round clipping from SliderWrapper overflow
```
The clipping kinda bugs me :/ It feels like a visual glitch, since it's not a gradual rounding animation, it's just being occluded by an invisible wrapper.

joshwcomeau commented on src/components/FeedbackButton/FeedbackButton.js:
```diff
@@ -0,0 +1,160 @@
+  height: 50px;
+  overflow: hidden;
+  border-radius: 25px;
+  user-select: none; /* diable text-selection */
```
Thanks for this comment :)

joshwcomeau commented on src/services/shell.service.js:
```diff
@@ -2,18 +2,49 @@
+    try {
+      return openedWindows[urlKey].show();
+    } catch (err) {
+      // swallow error - Window could be closed
```
It might be good to make a note of what the error is, exactly, when the window is closed (maybe see what `err.code` is?), and then to throw a new error if it's not what was expected.

I was bitten at work recently by a try/catch designed to catch a specific error, but it was swallowing a totally different error, and it made debugging really hard

AWolf81 replied:
Good catch. I'll add a handling. The error I'm swallowing here is `Object has been destroyed`. 
Not sure why I'm getting it because I'm already checking if the object is defined.

joshwcomeau commented on src/services/shell.service.js:
```diff
@@ -2,18 +2,49 @@
+
+  // Save Window so we can open it later
+  openedWindows[urlKey] = win;
+};
```
Hm so on my machine, it just opened a new chrome tab, the same way `shell.openExternal` works.

I think this is fine :) but yeah if you're keen to have it open a new app window, might need some debugging, at least on mac.

AWolf81 replied:
Oh, OK, for me this is working and opens a new electron window. Just tested on Windows.
So if we're using it later, we have to check the behaviour on Mac.

AWolf81 commented on src/components/DetectActive/DetectActive.js:
```diff
@@ -22,19 +29,26 @@ class DetectActive extends Component<Props, State> {
     return (
-      <span
+      <div
+        className={className}
```
It was used for the FeebackButton but I removed it so it was easier to create the animation. I'll have a look if I can use it. Usage of the `className` is like in this [Codesandbox](https://codesandbox.io/s/xjmqrzwj04).

AWolf81 commented:
@joshwcomeau Thanks for your review.

I've added everything. Also the animation is improved - the clipping happens now all the time and not just at the end of the animation. Disapearing of the text is still a bit fast but I think it's OK.

![screenrecording_feedbackbutton_latest](https://user-images.githubusercontent.com/3046542/49968753-0e171000-ff27-11e8-9ab6-bc172d84f68b.gif)

For the Github integration I missed that this is only available for a Pro account. So for now we don't have it but I think that's OK - we probably just have to copy some posts to Github issues and add a link in the Nolt post.

Would be great if you could add first posts to Nolt. So the board is not empty.
